### PR TITLE
Setup pipeline

### PR DESCRIPTION
**Pull request by NaomiVoerman**

### CODE REVIEW

File: run.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import time
+from scratch.nfold_analysis import nfold_analysis
```
Comment (rolfypma) [inline run.py]: klopt nog niet

File: rna/nfold_analysis.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    mle = MultiLabelEncoder(len(single_cell_types))
+
+    # ======= Load data =======
+    X_single, y_nhot_single, n_celltypes, n_features, n_per_celltype, label_encoder, present_markers, present_celltypes = get_data_per_cell_type(
+        single_cell_types=single_cell_types, markers=settings.markers)
+    y_single = mle.transform_single(mle.nhot_to_labels(y_nhot_single))
+    target_classes = string2vec(tc, label_encoder)
+
+    # ======= Initialize =======
+    train_accuracy, test_accuracy, mixture_accuracy, single_accuracy, cllr_menstr, cllr_menstr_mixt, cllr_nasal, \
```
Comment (rolfypma) [inline rna/nfold_analysis.py]: dit is een hele lelijke manier van uitkomsten bijhouden. kan dit in een datastructuur? bv een dict cllr met als keys de targetclasses?

File: rna/nfold_analysis.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+from rna.plotting import plot_histogram_log_lr, plot_boxplot_of_metric
+
+from scratch.single_analysis import get_accuracy
+
+# from lir.plotting import makeplot_hist_density
+
+BINARIZE = [True, False]
+LPS = [True, False]
+MODELS = ['MLP', 'MLR', 'XGB']
+# MODELS = ['MLR']
```
Comment (rolfypma) [inline rna/nfold_analysis.py]: weg

File: rna/nfold_analysis.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+from rna.analytics import augment_data, cllr, combine_samples
+from rna.constants import single_cell_types
+from rna.input_output import get_data_per_cell_type, read_mixture_data
+from rna.lr_system import MarginalMLPClassifier, MarginalMLRClassifier, MarginalXGBClassifier
+from rna.utils import vec2string, string2vec, MultiLabelEncoder
+from rna.plotting import plot_histogram_log_lr, plot_boxplot_of_metric
+
+from scratch.single_analysis import get_accuracy
+
+# from lir.plotting import makeplot_hist_density
```
Comment (rolfypma) [inline rna/nfold_analysis.py]: weg

File: rna/nfold_analysis.py (hunk 2 above)
Comment (rolfypma) [inline rna/nfold_analysis.py]: waarom zijn de cllrs per targetclass en de accuracies niet?

File: settings.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+Settings
+
+Options:
+    augment           If provided, use augmented data to train/calibrate the model with, otherwise use original data
+    binarize          If provided, make data binary, otherwise use original signal values
+    markers           If provided, include all markers, otherwise exclude 4 markers for control and gender
+    lps               If provided, use the label powerset method, otherwise use sigmoids
+    cal_probs         If proved, calibrate models on probabilities, otherwise on log10LRs)
+    nsamples          The number of augmented samples per combination: (N_SAMPLES_TRAIN, N_SAMPLES_CALIB, N_SAMPLES_TEST)
+    test_size         The size of the test data
```
Comment (rolfypma) [inline settings.py:11]: onduidelijk, is het nu 20% 50% 30% of 20% 40% 40% ?

File: settings.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+augment=True
+# binarize=True
+markers=True
+# lps=True
+cal_probs=False # Not incorporated in single_analysis.py
+nsamples=(50, 50, 25)
+test_size=0.2
+calibration_size=0.5
+# model='XGB'
```
Comment (rolfypma) [inline settings.py:24]: commented lines graag weg

Comment (rolfypma) [reply settings.py:24]: of beter, maak de nfold runbaar met deze settings, en zet er een laag overheen die opties als BINARIZE = [True, False] meegeeft, ipv hardcoden in nfold_analysis -> die is nu niet meer aan te roepen zonder een gridsearch te doen

File: rna/nfold_analysis.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+            for j, method in enumerate(LPS):
+                print("LPS: {}".format(method))
+
+                for k, MODEL in enumerate(MODELS):
+                    print("MODEL: {}".format(MODEL))
+                    model = model_with_correct_settings(MODEL, method)
+
+                    # ======= Calculate LRs before and after calibration =======
+                    if MODEL != 'MLR':
```
Comment (rolfypma) [inline rna/nfold_analysis.py:71]: dit is heel lelijk. beter bij het model aangeven of er gekalibreerd moet worden of niet. Op basis daarvan je data splitten. Nu krijgt MLR evenveel training data als de andere modellen, maar 0 kalibratie data? Dat zou toch niet moeten?

File: rna/nfold_analysis.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+                                         y_calib_nhot_augmented, X_test_augmented, X_mixtures, target_classes)
+
+                    else:
+                        X_calib_augmented_mlr = np.array([])
+                        y_calib_nhot_augmented_mlr = np.array([])
+                        lrs_before_calib, lrs_after_calib, lrs_before_calib_mixt, lrs_after_calib_mixt = \
+                            generate_lrs(model, mle, method, X_train_augmented_mlr, y_train_nhot_augmented_mlr, X_calib_augmented_mlr,
+                                         y_calib_nhot_augmented_mlr, X_test_augmented, X_mixtures, target_classes)
+
+                    plot_histogram_log_lr(lrs_before_calib, y_test_nhot_augmented, target_classes, label_encoder, savefig='scratch/hist_before_{}_{}_{}_{}'.format(n, binarize, method, MODEL))
```
Comment (rolfypma) [inline rna/nfold_analysis.py]: netter om os.path.join(scratch, ...) te gebruiken (OS onafhankelijk)

File: rna/nfold_analysis.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+                    # ======= Calculate accuracy and Cllr =======
+                    train_accuracy[n, i, j, k] = get_accuracy(model, mle, y_train_nhot_augmented, X_train_augmented, target_classes)
+                    test_accuracy[n, i, j, k] = get_accuracy(model, mle, y_test_nhot_augmented, X_test_augmented, target_classes)
+                    mixture_accuracy[n, i, j, k] = get_accuracy(model, mle, y_nhot_mixtures, X_mixtures, target_classes)
+                    if binarize == True:
+                        single_accuracy[n, i, j, k] = get_accuracy(model, mle, mle.inv_transform_single(y_test), X_test_bin, target_classes)
+                    else:
+                        single_accuracy[n, i, j, k] = get_accuracy(model, mle, mle.inv_transform_single(y_test), X_test_norm, target_classes)
+
+                    cllr_menstr[n, i, j, k] = cllr(lrs_after_calib[:, 0], y_test_nhot_augmented, target_classes[0])
```
Comment (rolfypma) [inline rna/nfold_analysis.py:100]: als dit geen losse variabelen zijn is dit hele code block een simpele loop

File: rna/nfold_analysis.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+    plot_boxplot_of_metric(cllr_skin_mixt, "cllr skin mixt", savefig='scratch/boxplot_cllr_skin_mixt')
+    plot_boxplot_of_metric(cllr_vag, "cllr vag", savefig='scratch/boxplot_cllr_vag')
+    plot_boxplot_of_metric(cllr_vag_mixt, "cllr vag_mixt", savefig='scratch/boxplot_cllr_vag_mixt')
+    plot_boxplot_of_metric(cllr_vag_menstr, "cllr vag menstr", savefig='scratch/boxplot_cllr_vag_menstr')
+    plot_boxplot_of_metric(cllr_vag_menstr_mixt, "cllr vag menstr mixt", savefig='scratch/boxplot_cllr_vag_menstr_mixt')
+
+    print("END")
+
+
+def model_with_correct_settings(model_no_settings, method):
```
Comment (rolfypma) [inline rna/nfold_analysis.py:180]: 'method' betekent hier of sigmoid gebruikt moet worden of lps? Dat moet er duidelijker staan (variabele naam en docstring)

File: rna/nfold_analysis.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+    elif model_no_settings == 'MLR' and method:
+        model = MarginalMLRClassifier(multi_class='multinomial', solver='newton-cg')
+    elif model_no_settings == 'MLR' and not method:
+        model = MarginalMLRClassifier()
+
+    elif model_no_settings == 'XGB' and method:
+        model = MarginalXGBClassifier()
+    elif model_no_settings == 'XGB' and not method:
+        model = MarginalXGBClassifier(method='sigmoid')
+
```
Comment (rolfypma) [inline rna/nfold_analysis.py:196]: else:
   raise Valuerror

File: rna/utils.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        self.nhot_of_combinations = make_nhot_matrix_of_combinations(n_classes)
+
+    def nhot_to_labels(self, y_nhot):
+        y = np.array([np.argwhere(np.all(self.nhot_of_combinations == y_nhot[i, :], axis=1)).flatten() for i in range(y_nhot.shape[0])])
+        return y.ravel()
+
+    def labels_to_nhot(self, y):
+        if len(y.shape) == 1 or y.shape[1] == 1:
+            n = y.shape[0]
+            y_nhot = np.vstack(self.nhot_of_combinations[y[i], :] for i in range(n))
```
Comment (rolfypma) [inline rna/utils.py:63]: dit genereert veel 
FutureWarning: arrays to stack must be passed as a "sequence" type such as list or tuple. Support for non-sequence iterables such as generators is deprecated as of NumPy 1.16 and will raise an error in the future.
  y_nhot = np.vstack(self.nhot_of_combinations[y[i], :] for i in range(n))

File: rna/lr_system.py [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+from lir import KDECalibrator
+from rna.analytics import get_mixture_columns_for_class
+
+from keras import Input, Model
+from keras.layers import Dense, Dropout
+
+
+class MarginalMLPClassifier():
+
+    def __init__(self, random_state=0, classifier=MLPClassifier, calibrator=KDECalibrator, activation='relu', MAX_LR=10):
```
Comment (rolfypma) [inline rna/lr_system.py:18]: het is raar dat MLPClassifier hier een optie is, terwijl deze in de klassenaam staat. Dus maak er of een grote generieke klasse van, of houd deze drie en haal de classifier parameter weg

File: rna/lr_system.py [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+    def fit_calibration(self, X, y_nhot, target_classes):
+        """
+        Makes calibrated model for each target class
+        """
+        lrs_per_target_class = self.predict_lrs(X, target_classes, with_calibration=False)
+
+        for i, target_class in enumerate(target_classes):
+            calibrator = self._calibrator()
+            loglrs = np.log10(lrs_per_target_class[:, i])
+            labels = np.max(np.multiply(y_nhot, target_class), axis=1)
```
Comment (rolfypma) [inline rna/lr_system.py:36]: hier mssn even commenten waarom een max werkt

File: rna/lr_system.py [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+            denominator = np.sum(prob[:, indices_of_non_target_class], axis=1)
+            lrs[:, i] = numerator/denominator
+
+        else: # sigmoid
+            # TODO: Incorporate priors
+            assert prob.shape[1] == target_classes.shape[0]
+            prob_target_class = prob[:, i].flatten()
+            prob_target_class = np.reshape(prob_target_class, -1, 1)
+            # if probability is exactly 1:
+            prob_target_class = np.where(prob_target_class == 1, 0.9999999999, prob_target_class)
```
Comment (rolfypma) [inline rna/lr_system.py:299]: dit gebeurt nu 2x? (onderaan de methode ook)

File: rna/nfold_analysis.py [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+                    else:
+                        X_calib_augmented_mlr = np.array([])
+                        y_calib_nhot_augmented_mlr = np.array([])
+                        lrs_before_calib, lrs_after_calib, lrs_before_calib_mixt, lrs_after_calib_mixt = \
+                            generate_lrs(model, mle, method, X_train_augmented_mlr, y_train_nhot_augmented_mlr, X_calib_augmented_mlr,
+                                         y_calib_nhot_augmented_mlr, X_test_augmented, X_mixtures, target_classes)
+
+                    plot_histogram_log_lr(lrs_before_calib, y_test_nhot_augmented, target_classes, label_encoder, savefig='scratch/hist_before_{}_{}_{}_{}'.format(n, binarize, method, MODEL))
+                    plot_histogram_log_lr(lrs_after_calib, y_test_nhot_augmented, target_classes, label_encoder, density=True, title='after', savefig='scratch/hist_after_{}_{}_{}_{}'.format(n, binarize, method, MODEL))
```
Comment (rolfypma) [inline rna/nfold_analysis.py:87]: waarom niet allebei density=True?

File: requirements.txt [hunk 17]
```diff
@@ -0,0 +1,6 @@
+matplotlib
+pandas
+numpy
+sklearn
+tqdm
+xlrd
```
Comment (rolfypma) [inline requirements.txt:6]: hier missen
xgboost
keras
tensorflow

File: rna/analytics.py [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+                    return False
+        # at least one of the target class should occur
+        if np.inner(binary, target_class)==0:
+            return False
+        return True
+
+    return [i for i in range(2 ** len(single_cell_types)) if binary_admissable(int_to_binary(i), target_class, priors)]
+
+
+# def classify_single(X, y_nhot):
```
Comment (rolfypma) [inline rna/analytics.py:180]: mag weg

File: rna/analytics.py [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+    """
+    try:
+        X = X[:, :-4]
+    except IndexError:
+        X = np.array([X[i][:, :-4] for i in range(X.shape[0])])
+
+    return X
+
+
+def cllr(lrs, y_nhot, target_class):
```
Comment (rolfypma) [inline rna/analytics.py:225]: zit dit niet in lir?

File: rna/analytics.py [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+#     plt.ylabel("Loss")
+#     plt.show()
+#
+#     # Compute confusion matrix
+#     cnf_matrix = confusion_matrix(y_test, y_pred)
+#     np.set_printoptions(precision=2)
+#     # print(cnf_matrix)
+
+
+def remove_markers(X):
```
Comment (rolfypma) [inline rna/analytics.py:213]: dit hoort zeker niet in analytics.py thuis

File: rna/constants.py [hunk 21]
```diff
@@ -0,0 +1,6 @@
+single_cell_types = \
+    ('Blood', 'Saliva', 'Vaginal.mucosa', 'Menstrual.secretion',
+     'Semen.fertile', 'Semen.sterile', 'Nasal.mucosa', 'Skin')
+
+marker_names = ['HBB', 'ALAS2', 'CD93', 'HTN3', 'STATH', 'BPIFA1', 'MUC4', 'MYOZ1', 'CYP2B7P1', 'MMP10', 'MMP7',
```
Comment (rolfypma) [inline rna/constants.py:5]: is het niet het meest zinnig om hier de laatste 4 weg te halen?

File: rna/single_analysis.py [hunk 22]
```diff
@@ -0,0 +1,282 @@
+"""
```
Comment (rolfypma) [inline rna/single_analysis.py:1]: willen we single_analysis.py houden? Lijkt extra code om te onderhouden zonder duidelijk doel?

**Review by rolfypma — APPROVED** [review]
kijk of je een logische mappenstructuur + indeling in modules kunt maken. aantal dingen van analytics kunnen wellicht naar utils, en wellicht eigen module maken voor het augmenteer gebeuren?

File: rna/analytics.py (hunk 20 above)
Comment (NaomiVoerman) [reply rna/analytics.py:213]: Verplaatst naar utils.py

File: rna/analytics.py (hunk 19 above)
Comment (NaomiVoerman) [reply rna/analytics.py:225]: Ja die gebruik ik, maar had nog niet de juiste input argumenten daarvoor en die 'maak' ik in deze functie.

File: rna/analytics.py (hunk 18 above)
Comment (NaomiVoerman) [reply rna/analytics.py:180]: Weggehaald

File: requirements.txt (hunk 17 above)
Comment (NaomiVoerman) [reply requirements.txt:6]: Toegevoegd

File: rna/lr_system.py (hunk 15 above)
Comment (NaomiVoerman) [reply rna/lr_system.py:299]: Ik kreeg soms een ZeroDivisionError vandaar dat ik dit had toegevoegd. Had niet gezien dat ik het inderdaad 2x doe nu. Heb die regel weggehaald.

File: rna/lr_system.py (hunk 14 above)
Comment (NaomiVoerman) [reply rna/lr_system.py:36]: Done

File: rna/nfold_analysis.py (hunk 11 above)
Comment (NaomiVoerman) [reply rna/nfold_analysis.py:196]: Done

File: rna/nfold_analysis.py (hunk 10 above)
Comment (NaomiVoerman) [reply rna/nfold_analysis.py:180]: Documentatie toegevoegd

File: rna/nfold_analysis.py (hunk 3 above)
Comment (NaomiVoerman) [reply rna/nfold_analysis.py]: Done

File: rna/nfold_analysis.py (hunk 4 above)
Comment (NaomiVoerman) [reply rna/nfold_analysis.py]: Done

File: rna/single_analysis.py (hunk 22 above)
Comment (NaomiVoerman) [reply rna/single_analysis.py:1]: Die kan inderdaad weg, aangezien de nfold_analysis.py hetzelfde gaat doen.

File: rna/nfold_analysis.py (hunk 16 above)
Comment (NaomiVoerman) [reply rna/nfold_analysis.py:87]: Gedaan

File: rna/nfold_analysis.py (hunk 8 above)
Comment (NaomiVoerman) [reply rna/nfold_analysis.py]: Zal ik aanpassen

File: settings.py (hunk 5 above)
Comment (NaomiVoerman) [reply settings.py:11]: Uitgelegd in documentatie

File: rna/lr_system.py (hunk 13 above)
Comment (NaomiVoerman) [reply rna/lr_system.py:18]: Aangepast

File: rna/nfold_analysis.py (hunk 2 above)
Comment (NaomiVoerman) [reply rna/nfold_analysis.py]: Aangepast

File: settings.py (hunk 6 above)
Comment (NaomiVoerman) [reply settings.py:24]: Aangepast

File: rna/nfold_analysis.py (hunk 7 above)
Comment (NaomiVoerman) [reply rna/nfold_analysis.py:71]: Aangepast

File: rna/nfold_analysis.py (hunk 9 above)
Comment (NaomiVoerman) [reply rna/nfold_analysis.py:100]: Aangepast

File: run.py (hunk 1 above)
Comment (NaomiVoerman) [reply run.py]: Nu wel verplaatst
